### PR TITLE
Fix DataFrame elemwise results in incorrect columns metadata

### LINKED ISSUE #696: Bug in DataFrame elemwise results in incorrect columns metadata
```
import numpy as np
import pandas as pd
import dask.dataframe as dd

df1 = pd.DataFrame(np.random.randn(5, 3), columns=list('abc'))
df2 = pd.DataFrame(np.random.randn(5, 3), columns=list('bcd'))

ddf1 = dd.from_pandas(df1, 2)
ddf2 = dd.from_pandas(df2, 2)

# NG, must be (a, b, c, d)
(ddf1 + ddf2).columns
# ('a', 'b', 'c')

(ddf1 + ddf2).compute()
#     a         b         c   d
# 0 NaN  1.820596  1.150497 NaN
# 1 NaN  2.290352 -0.490856 NaN
# 2 NaN -2.705724 -1.258149 NaN
# 3 NaN -0.694932 -2.714707 NaN
# 4 NaN  1.102297  1.512481 NaN
```

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #696. The fix is based on #695.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -1520,12 +1511,16 @@ def elemwise(op, *args, **kwargs):
...
-        return DataFrame(dsk, _name, columns, divisions)
-    else:
-        column_name = name or consistent_name(n for df in dfs
-                                              for n in df.columns)
-        return Series(dsk, _name, column_name, divisions)
+    if columns == no_default:
+        if len(dfs) >= 2 and len(dasks) != len(dfs):
+            # should not occur in current funcs
+            msg = 'elemwise with 2 or more DataFrames and Scalar is not supported'
+            raise NotImplementedError(msg)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Are there no element-wise operations that affect more than two operands?  What is the limitation here?

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -140,6 +138,17 @@ class _Frame(Base):
     _default_get = staticmethod(threaded.get)
     _finalize = staticmethod(finalize)
 
+    def __new__(cls, dsk, _name, metadata, divisions):
+        if isinstance(metadata, (Series, pd.Series)):
+            metadata = metadata.name
+        elif isinstance(metadata, (DataFrame, pd.DataFrame)):
+            metadata = metadata.columns
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Eventually it would be nice to document the constructor behavior somehow.  Perhaps not yet (things are changing rapidly) but we should consider this once things settle a bit.

File: dask/dataframe/utils.py [hunk 3]
```diff
@@ -160,3 +160,7 @@ def get_categories(df):
     if iscategorical(df.index.dtype):
         result['.index'] = df.index.categories
     return result
+
+
+def is_frame_or_series(x):
+    return isinstance(x, (pd.Series, pd.DataFrame))
```
Comment (mrocklin) [inline dask/dataframe/utils.py]: Why this change?

Comment (sinhrks) [inline dask/dataframe/utils.py]: This is for preciseness and simplification. ``NDFrame`` matches higher dimensional data structures like ``Panel`` (3D).

```
issubclass(pd.Panel, pd.core.generic.NDFrame)
True
```

Comment (mrocklin) [inline dask/dataframe/utils.py]: Why not inline the `isinstance` check directly?

    isinstance(df, pd.core.generic.NDFrame) -> isinstance(df, (pd.Series, pd.DataFrame))

File: dask/dataframe/core.py (hunk 1 above)
Comment (sinhrks) [inline dask/dataframe/core.py]: The limitation is caused by if ops requires 2 or more ``dask.DataFrame/Series`` and ``dask.core.Scalar``. Currently there is no operation affected by the limitation.

Explanation is:

- If arguments only contain one ``dask.DataFrame/dask.Series`` (and others are all scalars), metadata will not be changed. Thus, metadata can be predictable.
- If arguments only contain ``dask.DataFrame/dask.Series`` and no ``dask.core.Scalar``, we can apply op to empty-dummies. Thus, metadata can be predictable.
- If arguments contain 2 or more ``dask.DataFrame/dask.Series`` and 1 or more ``dask.core.Scalar``, we cannot perform empty op without knowing ``Scalar`` dtype. This case is currently not supported.

File: dask/dataframe/core.py (hunk 2 above)
Comment (sinhrks) [inline dask/dataframe/core.py]: Sure. Added the docstring.

File: dask/dataframe/utils.py (hunk 3 above)
Comment (sinhrks) [inline dask/dataframe/utils.py]: No mandatory reason. A background is ``utils`` already have shortening type check func like ``iscategorical``.

Changed to use ``isinstance``.
